### ISSUE TITLE
Developer Experience: `--no-config-lookup` allows for inline-config to be silently used

### ISSUE THREAD

**Issue by k0pernikus** [eslint/eslint#19232]
### Environment

Global ESLint version: 9.16.0

### What parser are you using?

@typescript-eslint/parser

### What did you do?

Developer Experience, eslint unexpected rules were applied during run with ` --no-config-lookup`. I also opened a [stackoverflow question](https://stackoverflow.com/q/79258162/457268) because of this behavior while debugging.

### What did you expect to happen?

**My expectations:**

I as a developer using `eslint` do not expect `eslint` to silently use inline-configs from my javascript files. If I add the option `--no-config-lookup`, I expect no config whatsoever to be looked up, neither a config file nor some inline config.

If you want to keep this behavior, I as a developer at least would expect the output of `eslint` to at least warn me that there are also now other rules being activated by my own code. 

As it was, it appeared like dark magic.

I was assuming some other configuration file was being used, that some parent file was used, I never expected my own JavaScript files to be the culprit. (My project is a legacy project. I want to, step by step, add eslint rules within the process.)

Furthermore, the scope of the inline config is strange. It seems to affect other files, also in my example two errors are reported, even though I only activated the rule after the first offender. The rules may then also be applied to other files as well.

### What actually happened?

**The problem:**

I wanted to run `eslint` with one and only one rule enabled.

I did it via:

```
eslint --no-config-lookup --rule "no-console: error" having_inline_rule_enabled.js  
```

Yet I saw different errors as well:

```
  1:1   error  Unexpected console statement  no-console
  1:13  error  Strings must use doublequote  quotes
  5:33  error  Strings must use doublequote  quotes

✖ 3 problems (3 errors, 0 warnings)
  2 errors and 0 warnings potentially fixable with the `--fix` option.
```

I was surprised that there were other rules checked for. I realized that I if one has inline rules within the javascript files, those are also active.

```
console.log('this should be reported');

/* eslint quotes: ["error", "double"], curly: 2 */

const thisShouldNotBeReported = 'this is unexpected';
```

I was only expecting the `console.log` to be reported. `--no-inline-config` solves my issue:

```
eslint --no-inline-config --no-config-lookup --rule "no-console: error" having_inline_rule_enabled.js  

/mnt/c/Users/philipp.kretzschmar/Documents/github/eslint_example/some.js
  1:1  error  Unexpected console statement  no-console

✖ 1 problem (1 error, 0 warnings)
```

### Link to Minimal Reproducible Example

https://gist.github.com/k0pernikus/50e41c8151e9bd183fbc19bede817332

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by fasttime** [comment]
Thanks for the issue @k0pernikus.
 
> I as a developer using `eslint` do not expect `eslint` to silently use inline-configs from my javascript files. If I add the option `--no-config-lookup`, I expect no config whatsoever to be looked up, neither a config file nor some inline config.

The current behavior is intended. As stated in [the documentation](https://eslint.org/docs/latest/use/command-line-interface#options), `--no-config-lookup` only disables looking for a config file (usually named `eslint.config.js`).

https://github.com/eslint/eslint/blob/e6b84f535e84050b0c63ae304eb17f9d181ac299/docs/src/use/command-line-interface.md?plain=1#L70

This option does not disable configuration supplied outside the config file, like inline config. As you already found out, inline config is disabled by the `--no-inline-config` option instead.
 
> If you want to keep this behavior, I as a developer at least would expect the output of `eslint` to at least warn me that there are also now other rules being activated by my own code.

There is currently no built-in way to be warned when a file being linted contains config comments, but `@eslint-community/eslint-plugin-eslint-comments` has a rule for that: https://eslint-community.github.io/eslint-plugin-eslint-comments/rules/no-use.html.

> As it was, it appeared like dark magic.
> 
> I was assuming some other configuration file was being used, that some parent file was used, I never expected my own JavaScript files to be the culprit. (My project is a legacy project. I want to, step by step, add eslint rules within the process.)

So, you were expecting files in your legacy project to contain problems but not to contain inline config comments? Or were you just unaware of the possibility to include config comments in code?

> Furthermore, the scope of the inline config is strange. It seems to affect other files, also in my example two errors are reported, even though I only activated the rule after the first offender. The rules may then also be applied to other files as well.

Inline config comments should not affect other files. If you have an example where this would happen, please show it so we can have a look.

The fact that in your example file two errors are reported is expected is expected, because `/* eslint */` config comments, unlike other config comments, always apply to the whole file regardless of where they occur.

**Comment by k0pernikus** [comment]
> > As it was, it appeared like dark magic.
> > I was assuming some other configuration file was being used, that some parent file was used, I never expected my own JavaScript files to be the culprit. (My project is a legacy project. I want to, step by step, add eslint rules within the process.)
> 
> So, you were expecting files in your legacy project to contain problems but not to contain inline config comments? Or were you just unaware of the possibility to include config comments in code?

Both.

The legacy project has thousands of issues. This is why I want to start on a very low level with adding checks. (I wanted to use not use a config file, to have a quick one-liner that I can use for a git-hook.)

I expect for linting tools to provide the feature to DISABLE rules within files on the fly. I do not expect the option to ENABLE rules on the fly for a file.

I am OK with the current behavior being the intended behavior. I am not OK with `eslint` being silent about it.

Esp. when I add flags like `-vvv` or `--print-config` I expect to figure out where the rules are being activated.

Esp. `print-config` also confuses me (might only make sense when having a config file?):

```
eslint --print-config --no-config-lookup **.js
Value for 'print-config' of type 'path::String' required.
You're using eslint.config.js, some command line flags are no longer available. Please see https://eslint.org/docs/latest/use/command-line-interface for details.
```

I don't use any `eslint.config.js`.

> > Furthermore, the scope of the inline config is strange. It seems to affect other files, also in my example two errors are reported, even though I only activated the rule after the first offender. The rules may then also be applied to other files as well.
> 
> Inline config comments should not affect other files. If you have an example where this would happen, please show it so we can have a look.
> 
> The fact that in your example file two errors are reported is expected is expected, because `/* eslint */` config comments, unlike other config comments, always apply to the whole file regardless of where they occur.

That they applied to the whole files, I find unexpected. I expect the rules to only be applied once they are declared.

I assume that some rules are applied incorrectly to different files, as I tried checking just single files that were marked as having errors when all files were being checked, but not once I just had a singular check.

I tried to create a reproducible example, but was yet not able too. I update if I actually find such an example.

**Comment by fasttime** [comment]
Thanks for the update!

> Both.
> 
> The legacy project has thousands of issues. This is why I want to start on a very low level with adding checks. (I don't want to use a config file in order to have a quick one-liner that I can use for a git-hook.)

That's a reasonable point. We are implementing a change that will allow users to suppress existing lint problems so that it will be easier to adopt ESLint progressively. You may want to follow #19159.

> I expect for linting tools to provide the feature to DISABLE rules within files on the fly. I do not expect the option to ENABLE rules on the fly for a file.

Enabling rules per inline config has been a feature of ESLint for a long time. It cannot be removed, but it can be disabled with `--no-inline-config` from the CLI.

> I am OK with the current behavior being the intended behavior. I am not OK with `eslint` being silent about it.
> 
> Esp. when I add flags like `-vvv` or `--print-config` I expect to figure out where the rules are being activated.
> 
> Esp. `print-config` also confuses me (might only make sense when having a config file?):
> 
> ```
> eslint --print-config --no-config-lookup **.js
> Value for 'print-config' of type 'path::String' required.
> You're using eslint.config.js, some command line flags are no longer available. Please see https://eslint.org/docs/latest/use/command-line-interface for details.
> ```
> 
> I don't use any `eslint.config.js`.

Note that there is no flag named `-vvv`. As noted in [the documentation](https://eslint.org/docs/latest/use/command-line-interface#--print-config), `--print-config` must be used with the path of a config file. Unfortunately, there aren't any great tools to help debugging inline configs.

> I assume that some rules are applied incorrectly to different files, as I tried checking just single files that were marked as having errors when all files were being checked, but not once I just had a singular check.
> 
> I tried to create a reproducible example, but was yet not able too. I update if I actually find such an example.

That could be a bug, be we'll need a repro to understand what is going on.

**Comment by k0pernikus** [comment]
For completeness' sake: `--debug` provides a hint that some other rules are loaded, even though it is not made explicit that they are set via infile comment config, yet that can be inferred via the order of the output.

```
  ...
  eslint:rules Loading rule 'no-console' (remaining=289) +0ms
  ...
  eslint:languages:js Scope analysis successful: /eslint_example/some.js +1ms
  eslint:rules Loading rule 'quotes' (remaining=288) +128ms
  eslint:rules Loading rule 'curly' (remaining=287) +12ms
  ...
```

At least this provides some meaningful context which files enable which rules.

**Comment by fasttime** [comment]
I'm going to close this issue because the described behavior of the `--no-config-lookup` is intended and not a bug. Please, feel free to open a change request if you would like to propose improving the current behavior or adding a warning. If you observe that a rule is being incorrectly reported on a different file, please file a bug with a reproducible example. To ask for clarifications about other aspects like the output when the `--debug` flag is used, you can open a [discussion](https://github.com/eslint/eslint/discussions).
